upstairs/src/lib.rs

```diff
@@ -2313,20 +2455,14 @@ async fn up_ds_listen(up: &Arc<Upstairs>, mut ds_done_rx: mpsc::Receiver<u64>) {
             let ds_id = done.ds_id;
             assert_eq!(*ds_id_done, ds_id);
 
-            // Move this work to Acked
-            let mut ack_ready = false;
-            for cid in 0..3 {
-                if let Some(IOState::AckReady) = done.state.get(&cid) {
-                    done.state.insert(cid, IOState::Acked);
-                    ack_ready = true;
-                    break;
-                }
-            }
-            // Verify we did find an AckReady downstairs IO
-            assert!(ack_ready);
             let data = done.data.take();
-            gw.ds_complete(gw_id, ds_id, data);
-            crutrace_work_done(&done.work, gw_id);
+
+            // Verify we did find an AckReady downstairs IO
+            assert!(work.ack(ds_id));
+
+            gw.ds_complete(gw_id, ds_id, data, work.result(ds_id));
```

Comment:
I don't think this will work (unless I've got this all backawards).  This loop is walking a list of ds_id's built in the ackable_work() method.  That method only searches for an `IOState::AckReady`, and won't find jobs where there a are either all errors, or a mix of errors and done but not enough to return success to the guest.

I think perhaps we need to update ackable_work to return jobs that:
 * Have enough to return success (and data if read) to the guest.
 * Have enough answers to return failure to the guest.  In this case 2 failures from a write/flush would be enough to return error to the guest, while a read requires all three to error.

It's a combo of the current `ackable_work()` and the `result` method, sort of.

And, we might want to consider an "AckError" state, to signify which IO we chose to return error on.
Maybe an ErrReady state, and that becomes AckError?  Then the `ackable_work()` method could check those two states and return the list.

Reply:
Yeah, that looks good